src/compile/data/formatparse.ts

```diff
@@ -52,77 +53,85 @@ export class ParseNode extends DataFlowNode {
     this._parse = parse;
   }
 
-  public static make(parent: DataFlowNode, model: Model) {
-    const parse = {};
-    const calcFieldMap = {};
-
-    (model.transforms || []).forEach((transform: Transform) => {
-      if (isCalculate(transform)) {
-        calcFieldMap[transform.as] = true;
-      } else if (isFilter(transform)) {
-        forEachLeaf(transform.filter, (filter) => {
-          if (isFieldPredicate(filter)) {
-            if (filter.timeUnit) {
-              parse[filter.field] = 'date';
-            }
-          }
-        });
-      }
-    }, {});
+  /**
+   * Creates a parse node from a data.format.parse and updates ancestorParse.
+   */
+  public static makeExplicit(parent: DataFlowNode, model: Model, ancestorParse: AncestorParse) {
+    // Custom parse
+    let explicit = {};
+    const data = model.data;
+    if (data && data.format && data.format.parse) {
+      explicit = data.format.parse;
+    }
+
+    return this.makeWithAncestors(parent, explicit, {}, ancestorParse);
+  }
+
+  /**
+   * Creates a parse node for implicit parsing from a model and updates ancestorParse.
+   */
+  public static makeImplicit(parent: DataFlowNode, model: Model, ancestorParse: AncestorParse) {
+    const implicit = {};
 
     if (isUnitModel(model) || isFacetModel(model)) {
       // Parse encoded fields
       model.forEachFieldDef(fieldDef => {
         if (isTimeFieldDef(fieldDef)) {
-          parse[fieldDef.field] = 'date';
+          implicit[fieldDef.field] = 'date';
         } else if (isNumberFieldDef(fieldDef)) {
-          if (calcFieldMap[fieldDef.field] || isCountingAggregateOp(fieldDef.aggregate)) {
+          if (isCountingAggregateOp(fieldDef.aggregate)) {
             return;
           }
-          parse[fieldDef.field] = 'number';
+          implicit[fieldDef.field] = 'number';
         } else if (accessPathDepth(fieldDef.field) > 1) {
           // For non-date/non-number (strings and booleans), derive a flattened field for a referenced nested field.
           // (Parsing numbers / dates already flattens numeric and temporal fields.)
-          parse[fieldDef.field] = 'flatten';
+          implicit[fieldDef.field] = 'flatten';
         }
       });
     }
 
-    // Custom parse should override inferred parse
-    const data = model.data;
-    if (data && data.format) {
-      const p = data.format.parse;
-      if (p === null) {
-        return null;  // disable parsing completely
-      } else if (p) {
-        keys(p).forEach(field => {
-          parse[field] = p[field];
-        });
+    return this.makeWithAncestors(parent, {}, implicit, ancestorParse);
+  }
+
+  /**
+   * Creates a parse node from two sets of explicit and implicit parse and updates ancestorParse.
+   */
+  public static makeWithAncestors(parent: DataFlowNode, explicit: Dict<string>, implicit: Dict<string>, ancestorParse: AncestorParse) {
```

Comment:
Why do you need to pass `ancestorParse` here?   Shouldn't it be a part of `parent`?  

(Sorry this is very complicated so it's hard to get why.)

Reply:
No, the ancestor parse applies to the current model as well. I could simplify this and not make a copy, I guess.

Reply:
If you can, better do it now.

Reply:
Actually, that's incorrect when we have forks because every subtree needs to track parsed fields separately.

Reply:
I don't understand your explanation here yet... 

Please explain "Why do you need to pass ancestorParse here?"

Reply:
Oh, my response was to making a copy.

`ancestorParse` is currently not in nodes but passed as a separate object. I was thinking about adding this information to the nodes themselves but that makes a lot of the merging logic more complicated. I created an issue for this in https://github.com/vega/vega-lite/issues/3715